api/graphite.go

```diff
@@ -59,44 +59,61 @@ func (s *Server) findSeries(ctx context.Context, orgId int, patterns []string, s
 		return nil, err
 	}
 	log.Debug("HTTP findSeries for %v across %d instances", patterns, len(peers))
-	errors := make([]error, 0)
-	series := make([]Series, 0)
-
-	var mu sync.Mutex
 	var wg sync.WaitGroup
+
+	responses := make(chan struct {
+		series []Series
+		err    error
+	}, 1)
+	findCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
 	for _, peer := range peers {
 		log.Debug("HTTP findSeries getting results from %s", peer.Name)
 		wg.Add(1)
 		if peer.IsLocal() {
 			go func() {
-				result, err := s.findSeriesLocal(ctx, orgId, patterns, seenAfter)
-				mu.Lock()
+				result, err := s.findSeriesLocal(findCtx, orgId, patterns, seenAfter)
 				if err != nil {
-					errors = append(errors, err)
+					// cancel requests on all other peers.
+					cancel()
 				}
-				series = append(series, result...)
-				mu.Unlock()
+				responses <- struct {
+					series []Series
+					err    error
+				}{result, err}
 				wg.Done()
 			}()
 		} else {
 			go func(peer cluster.Node) {
-				result, err := s.findSeriesRemote(ctx, orgId, patterns, seenAfter, peer)
-				mu.Lock()
+				result, err := s.findSeriesRemote(findCtx, orgId, patterns, seenAfter, peer)
 				if err != nil {
-					errors = append(errors, err)
+					// cancel requests on all other peers.
+					cancel()
 				}
-				series = append(series, result...)
-				mu.Unlock()
+				responses <- struct {
+					series []Series
+					err    error
+				}{result, err}
 				wg.Done()
 			}(peer)
 		}
 	}
-	wg.Wait()
-	if len(errors) > 0 {
-		err = errors[0]
+
+	// wait for all findSeries goroutines to end, then close our responses channel
+	go func() {
+		wg.Wait()
```

Comment:
it seems like creating a goroutine for this is not necessary. just `wg.Wait()` and `close(responses)`, without creating a routine, would have the same effect